### PR TITLE
Relevacny builder module

### PR DESCRIPTION
**Pull request by mfranz1**
Story: https://app.clubhouse.io/clarkcan/story/4564/build-relevancy-builder

This PR completes the addition of the Relevancy Builder to the clark-client. The builder is accessible through the admin dashboard in the meatball options of a learning object. On the admin dashboard, the learning objects are able to be filtered by topics. The relevancy builder allows admins and editors to tag a learning object with unique topics and map the objects outcomes to standard-guidelines. 

image
image
image
image
image

### CODE REVIEW

File: src/app/onion/learning-object-builder/learning-object-builder.component.html [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <div class="learning-object-builder-wrapper" [@builderTransition]="getState(o)">
   <!-- The ngIf directive on onion-builder-navbar forces the :enter event to fire for router animations and should NOT be removed -->
-  <onion-builder-navbar *ngIf="true" [adminMode]="adminMode"></onion-builder-navbar>
```
Comment (pzalep1) [inline src/app/onion/learning-object-builder/learning-object-builder.component.html:3]: This should probably stay.

File: src/app/admin/components/learning-object-list-item/learning-object-list-item.component.html [hunk 2]
```diff
@@ -37,6 +37,7 @@
       <li (click)="toggleRelevancyDate(true)"><i class="fal fa-history"></i>Mark Next Check Date</li>
       <li *ngIf="hasParents === false" (click)="toggleChangeAuthorModal(true)"><i class="far fa-user"></i>Change Author</li>
       <li *ngIf="learningObject.status === 'released'" (click)="toggleUnreleaseConfirm(true)"><i class="far fa-eye-slash"></i>Unrelease</li>
+      <li [routerLink]="['/onion/relevancy-builder', learningObject.id]"><i class="far fa-puzzle-piece"></i>Object Relevancy</li>
```
Comment (pzalep1) [inline src/app/admin/components/learning-object-list-item/learning-object-list-item.component.html:40]: Can we say Map and Tag?

File: src/app/onion/relevancy-builder/builder-store.service.ts [hunk 3]
```diff
@@ -0,0 +1,686 @@
...
+  CREATE_OUTCOME,
+  FETCH_OBJECT,
+  FETCH_OBJECT_MATERIALS,
+  UPDATE_FILE_DESCRIPTION,
+  UPDATE_OBJECT,
+  UPDATE_OUTCOME,
+  SUBMIT_REVIEW,
+  CANCEL_SUBMISSION,
+  DELETE_OUTCOME,
+  ADD_FILE_META,
```
Comment (pzalep1) [inline src/app/onion/relevancy-builder/builder-store.service.ts:56]: I think we can get rid of all the stuff that isn't used in this builder.

File: src/app/onion/relevancy-builder/components/builder-navbar/builder-navbar.component.html [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      ></clark-lo-status-indicator>
+      <!-- learning object and new tab  preview go here -->
+    </div>
+    <div class="top__right">
+      <a (activate)="historySnapshot.rewind('/admin/learning-objects')">Cancel</a>
+    </div>
+  </div>
+  <div class="builder-navbar-wrapper__bottom">
+    <div class="builder-navbar-wrapper__bottom-left-menu">
+      <!-- update routerLinkActive with class used to show that link is active -->
```
Comment (pzalep1) [inline src/app/onion/relevancy-builder/components/builder-navbar/builder-navbar.component.html:22]: Remove commented code.

File: src/app/onion/relevancy-builder/components/builder-navbar/builder-navbar.component.ts [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+      });
+    this.historySnapshot = this.history.snapshot();
+  }
+
+  /**
+   * Build the states map for the status tooltips and icons
+   *
+   * @memberof BuilderNavbarComponent
+   */
+  buildTooltip() {
```
Comment (pzalep1) [inline src/app/onion/relevancy-builder/components/builder-navbar/builder-navbar.component.ts:70]: Do we still need this function?

File: src/app/onion/relevancy-builder/components/outcome/outcome-typeahead/outcome-typeahead.component.html [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        {{ verb | titlecase }}
+      </div>
+      <input [ngClass]="{'withVerb': verb}" placeholder="Outcome text" type="text"  [value]="text" />
+    </div>
+  </div>
+
+  <!-- 
+
+    // little arrow thing
+    <span *ngIf="goodVerb">
```
Comment (pzalep1) [inline src/app/onion/relevancy-builder/components/outcome/outcome-typeahead/outcome-typeahead.component.html:13]: Remove commented code.

File: src/app/onion/relevancy-builder/components/scafold/scafold.component.html [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+<div class="header">
+    <div class="title">
+      Learning Object Topics
+    </div>
+</div>
+<ng-template [ngIf]="topics && topics.length > 0" class="learning-object-topics">
+  <div class="pills" >
+    <div *ngFor="let t of topics; let i = index" class="pill"> 
+      <span (activate)="toggleTopics(t)" [ngClass]="{'tagged': tagged}">{{ t }}</span>
+      <!-- <div class = "info">
```
Comment (pzalep1) [inline src/app/onion/relevancy-builder/components/scafold/scafold.component.html:11]: Remove commented code.

File: src/app/onion/relevancy-builder/relevancy-builder.component.ts [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+        // if name parameter found, instruct store to fetch full learning object
+        if (revision !== undefined && id) {
+          this.isRevision = true;
+          this.store.isRevision = true;
+          this.store.fetch(id, revision, authorUsername).then(learningObject => {
+            this.setBuilderMode(learningObject);
+          });
+        } else if (id) {
+          this.store.fetch(id).then(learningObject => {
+            if (learningObject.status === LearningObject.Status.RELEASED) {
```
Comment (pzalep1) [inline src/app/onion/relevancy-builder/relevancy-builder.component.ts:146]: All learning objects that are not unreleased can be loaded in here so we should make sure we're not throwing people out of the builder.

File: src/app/onion/relevancy-builder/relevancy-builder.component.ts [hunk 9]
```diff
@@ -0,0 +1,322 @@
+import { takeUntil } from 'rxjs/operators';
```
Comment (pzalep1) [inline src/app/onion/relevancy-builder/relevancy-builder.component.ts:1]: All unecessary code should be taken out of this file as well.

File: src/app/onion/relevancy-builder/relevancy-builder.routing.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+import { Routes, RouterModule } from '@angular/router';
+
+import { OutcomePageComponent } from './pages/outcome-page/outcome-page.component';
+import { RelevancyBuilderComponent } from './relevancy-builder.component';
+
+const builder_routes: Routes = [
+    { path: '', component: RelevancyBuilderComponent,
+    children: [
+      { path: '', redirectTo: 'mappings', pathMatch: 'full' },
+      { path: 'mappings', component: OutcomePageComponent, data: { state: 'outcomes' } },
```
Comment (pzalep1) [inline src/app/onion/relevancy-builder/relevancy-builder.routing.ts:11]: What is this route for?

File: src/app/onion/relevancy-builder/validators/learning-object.validator.ts [hunk 11]
```diff
@@ -0,0 +1,301 @@
+import { Injectable } from '@angular/core';
```
Comment (pzalep1) [inline src/app/onion/relevancy-builder/validators/learning-object.validator.ts:1]: All unnecessary code should be removed from this file as well.

**Review by pzalep1 — CHANGES_REQUESTED** [review]

File: src/app/onion/learning-object-builder/learning-object-builder.component.html (hunk 1 above)
Comment (mfranz1) [reply src/app/onion/learning-object-builder/learning-object-builder.component.html:3]: Good catch, my bad

**Review by Cwagne17 — APPROVED** [review]

File: src/app/admin/components/filter-search/filter-search.component.ts [hunk 12]
```diff
@@ -22,23 +24,29 @@ import { ToastrOvenService } from 'app/shared/modules/toaster/notification.servi
...
 
   private _selectedCollection: Collection;
 
+  private _selectedTopic: Topic;
+
   @Input() adminOrEditor: boolean;
   @Input() showStatus: boolean;
   @Output() statusFilter = new EventEmitter<any[]>();
   @Output() collectionFilter = new EventEmitter<string>();
+  @Output() topicFilter = new EventEmitter<any[]>();
```
Comment (EVogel1999) [inline src/app/admin/components/filter-search/filter-search.component.ts:42]: Lets add a type to this

File: src/app/core/relevancy.service.ts [hunk 13]
```diff
@@ -48,6 +45,82 @@ export class RelevancyService {
...
+            resolve(res);
+          },
+          (err) => {
+            reject(err);
+          }
+        );
+    });
+  }
+
+  async updateObjectTopics(username: string, id: string, topicIds: string[]): Promise<void> {
```
Comment (EVogel1999) [inline src/app/core/relevancy.service.ts:78]: These should have block comments.

File: src/app/onion/relevancy-builder/components/outcome/outcome.component.scss [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+    top: 0;
+    bottom: 0;
+    left: -6px;
+    background: $light-blue;
+    width: 5px;
+    transition: all 0.2s ease;
+  }
+
+  &.active {
+    background: #f8f9fd;
```
Comment (EVogel1999) [inline src/app/onion/relevancy-builder/components/outcome/outcome.component.scss:31]: This should be a standard color.

File: src/app/onion/relevancy-builder/components/standard-outcomes/outcome-list-item/outcome-list-item.component.scss [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+      color: $gold;
+      margin-right: 6px;
+      font-size: $normal;
+    }
+  }
+
+  &.selected {
+    background: $light-blue-gradient;
+    color: rgba(255, 255, 255, 0.8) !important;
+    box-shadow: inset 0 2px 6px 2px rgba(0, 0, 0, 0.1);
```
Comment (EVogel1999) [inline src/app/onion/relevancy-builder/components/standard-outcomes/outcome-list-item/outcome-list-item.component.scss:39]: The colors in this file should be standard colors.

**Review by EVogel1999 — CHANGES_REQUESTED** [review]
Just a few comments about styling.  Also just going to put what we talked about this morning in here:

- Fix the standard search to auto-suggest when an outcome is selected

**Review by jouypo23 — APPROVED** [review]

**Review by pzalep1 — COMMENTED** [review]
Just that one addition to the cards and this should be good!

**Review by pzalep1 — APPROVED** [review]
MERGE!!!

**Pull request merged**